osbuild/buildroot.py

```diff
@@ -14,115 +19,169 @@
+        self._vardir = var
+        self._libdir = libdir
+        self._runner = runner
+
+    @staticmethod
+    def _mknod(path, name, mode, major, minor):
+        os.mknod(os.path.join(path, name),
+                 mode=(stat.S_IMODE(mode) | stat.S_IFCHR),
+                 device=os.makedev(major, minor))
+
+    def __enter__(self):
+        self._exitstack = contextlib.ExitStack()
+        with self._exitstack:
+            # We create almost everything directly in the container as temporary
+            # directories and mounts. However, for some things we need external
+            # setup. For these, we create temporary directories which are then
+            # bind-mounted into the container.
+            #
+            # For now, this includes:
+            #
+            #   * We create an API directory where the caller can place sockets
+            #     before we bind-mount it into the container on
+            #     `/run/osbuild/api`.
+            #
+            #   * We create a tmpfs instance *without* `nodev` which we then use
+            #     as `/dev` in the container. This is required for the container
+            #     to create device nodes for loop-devices.
+            #
+            #   * We create a temporary directory for variable data and then use
+            #     it as '/var' in the container. This allows the container to
+            #     create throw-away data that it does not want to put into a
+            #     tmpfs.
+
+            api = tempfile.TemporaryDirectory(prefix="osbuild-api-", dir=self._rundir)
+            self.api = self._exitstack.enter_context(api)
+
+            dev = tempfile.TemporaryDirectory(prefix="osbuild-dev-", dir=self._rundir)
+            self.dev = self._exitstack.enter_context(dev)
+
+            var = tempfile.TemporaryDirectory(prefix="osbuild-var-", dir=self._vardir)
+            self.var = self._exitstack.enter_context(var)
+
+            subprocess.run(["mount", "-t", "tmpfs", "-o", "nosuid", "none", self.dev], check=True)
+            self._exitstack.callback(lambda: subprocess.run(["umount", "--lazy", self.dev], check=True))
+
+            self._mknod(self.dev, "full", 0o666, 1, 7)
+            self._mknod(self.dev, "null", 0o666, 1, 3)
+            self._mknod(self.dev, "random", 0o666, 1, 8)
+            self._mknod(self.dev, "urandom", 0o666, 1, 9)
+            self._mknod(self.dev, "tty", 0o666, 5, 0)
+            self._mknod(self.dev, "zero", 0o666, 1, 5)
+
+            self._exitstack = self._exitstack.pop_all()
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        with self._exitstack:
+            pass # Cleanup of the `ExitStack`.
 
     def run(self, argv, binds=None, readonly_binds=None):
         """Runs a command in the buildroot.
 
-        Its arguments mean the same as those for subprocess.run().
+        Takes the command and arguments, as well as bind mounts to mirror
+        in the build-root for this command.
```

Comment:
Add a `Can only be called with an active context` or `.. when the context is active`?